internal/monitors/jaegergrpc/monitor.go

```diff
@@ -0,0 +1,161 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type _sourceKey int
+
+const gracefulShutdownTimeout = time.Second * 5
+
+var (
+	sourceKey _sourceKey
+	logger    = utils.NewThrottledLogger(log.WithFields(log.Fields{"monitorType": monitorType}), 30*time.Second)
+)
+
+func init() {
+	monitors.Register(&monitorMetadata, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+// TLSCreds is a configuration struct for specifying a cert file and key file for tls
+type TLSCreds struct {
+	// The cert file to use for tls
+	CertFile string `yaml:"certFile"`
+	// The key file to use for tls
+	KeyFile string `yaml:"keyFile"`
+}
+
+func (tls *TLSCreds) Credentials() (credentials.TransportCredentials, error) {
+	var creds credentials.TransportCredentials
+	var err error
+	if tls != nil {
+		creds, err = credentials.NewServerTLSFromFile(tls.CertFile, tls.KeyFile)
+	}
+
+	return creds, err
+}
+
+// Config for this monitor
+type Config struct {
+	config.MonitorConfig `yaml:",inline" acceptsEndpoints:"false" singleInstance:"true"`
+	// The host:port on which to listen for datapoints.  The listening server
+	// accepts datapoints on the same HTTP path that ingest/gateway accepts
+	// them (e.g. `/v2/datapoint`, `/v1/trace`).  Requests to other paths will
+	// return 404s.
+	ListenAddress string `yaml:"listenAddress" default:"0.0.0.0:14250"`
+	// TLS are optional tls credential settings to configure the GRPC server with
+	TLS *TLSCreds `yaml:"tls, omitempty"`
+	// HTTP timeout duration for both read and writes. This should be a
+	// duration string that is accepted by https://golang.org/pkg/time/#ParseDuration
+	ServerTimeout time.Duration `yaml:"serverTimeout" default:"5s"`
+}
+
+// Monitor that accepts and forwards SignalFx data
+type Monitor struct {
+	Output types.Output
+	lock   sync.Mutex
+	grpc   *grpc.Server
+	ln     net.Listener
+}
+
+var _ api_v2.CollectorServiceServer = (*Monitor)(nil)
+
+// PostSpans implements the jeager api_v2.CollectorServiceServer interface.  The grpc server will pass the jaeger
+// batches it receives to this method.  This method will convert the jaeger batches to SignalFx spans and pass them
+// on to the output writer.
+func (m *Monitor) PostSpans(ctx context.Context, r *api_v2.PostSpansRequest) (*api_v2.PostSpansResponse, error) {
+	if r != nil {
+		// convert the batch to SignalFx metrics
+		spans := jaegerprotobuf.JaegerProtoBatchToSFX(&r.Batch)
+
+		// tag the source on the span meta data
+		source, hasSource := ctx.Value(sourceKey).(net.IP)
+		for i := range spans {
+			if hasSource {
+				if spans[i].Meta == nil {
+					spans[i].Meta = map[interface{}]interface{}{}
+				}
+				spans[i].Meta[constants.DataSourceIPKey] = source
+			}
+		}
+
+		// send the spans on through the agent
+		m.Output.SendSpans(spans...)
+	}
+
+	return &api_v2.PostSpansResponse{}, nil
+}
+
+// Configure the monitor and kick off volume metric syncing
+func (m *Monitor) Configure(conf *Config) error {
+	m.lock.Lock()
```

Comment:
You don't need a lock because `Configure` can now only be called once per monitor instance.

Reply:
Also `Configure` and `Shutdown` happen in the same goroutine so there won't be any issues of memory coherency.